### ISSUE TITLE
onRequest is not triggered when calling action via sendRequest

### ISSUE THREAD

**Issue by megawac** [klis87/redux-saga-requests#236]
I'm using `sendRequest` in a couple spots to load several resources in parallel and yield for their completion. Unfortunately I noticed when I try to use `sendRequest` the `onRequest` handler in the `reducers` fails to be triggered as the base [action never gets triggered](https://github.com/klis87/redux-saga-requests/blob/db88e1d8b807692f146f3c19c773518e36c99938/packages/redux-saga-requests/src/reducers.js#L273). This causes pending states to not get tracked correctly and I've been getting pending states in the negatives as I fetch other resources.

This is a simplified version of what I'm doing (

```js
const getFeaturesAction = createAction(`get feature`, ({ mapId, type }) => {
  return {
    request: {
      url: `/v2/maps/${type}/?map=${mapId}`
    },
  };
});

function createFeatureFetchAction(type) {
  return createAction(`get feature of ${type}`, ({  }) => {
    return {
      request: {
        url: `/v2/maps/${type}/?map=${mapId}`
      },
    };
  });
}

const fetchFeature1 = createFeatureFetchAction('feature-set-1');
const fetchFeature2 = createFeatureFetchAction('feature-set-2');
const fetchFeature3 = createFeatureFetchAction('feature-set-3');

function* getFeaturesSaga({ mapId }) {
  yield all([
    call(sendRequest,fetchFeature1({ mapId })),
    call(sendRequest, fetchFeature2({ mapId })),
    call(sendRequest, fetchFeature3({ mapId })),
  ])

  // Do some stuff....
}

const reducer = requestsReducer({
  type: fetchFeature1
});

// reducer2, 3...
```

Perhaps there is another way you'd suggest achieving this workflow instead?

**Comment by klis87** [comment]
Check docs of `sendRequest` https://github.com/klis87/redux-saga-requests#sendrequest , from my experience usually sagas react of already dispatched request actions, so by default `sendRequest` doesnt dispatch request, see `dispatchRequestAction` in docs, which u can change to `true`

Alternatively just use
```js
 yield all([
    put(fetchFeature1({ mapId })),
    put(fetchFeature2({ mapId })),
    put(fetchFeature3({ mapId })),
  ])
```
which will be then handled by `watchRequests`, Generally `sendRequest` can be used if you wanna control which interceptor should be ignored etc, for simple cases dispatching request actions is all you need. `sendRequest` is also nice if u need to see response or error in your saga. But even for this it is not necessary, you can use `promiseMiddleware` with `yield put.resolve(requestAction)` instead.

**Comment by megawac** [comment]
Actually I think my issue was I had `resetOn: [fetchFeature1]` which caused that part of the switch statement to never get hit as the reducer would see the `fetchFeature1` action, see that it should `resetOn` that type of action and `return`.

Would you be open to me changing [this line](https://github.com/klis87/redux-saga-requests/blob/master/packages/redux-saga-requests/src/reducers.js#L116) to:

```js
nextState = {
      ...getInitialState(state, reducer, config),
      pending: state.pending,
};
```

**Comment by klis87** [comment]
I guess you use 2nd param of `requestReducer` to merge your state with request state?

**Comment by megawac** [comment]
In my use case when I go to fetch a resource (a map) I load the assets of the map which are 10 or so files in parallel. When I dispatch my request action type for each of those assets I want to reset my reducers states while the fetches are in progress and as the assets load they will appear in my interface. The reason I want to reset is that I want the assets to load into on demand but I don't want them to be visible for the previous map.

So my strategy was to have my reducers set up like `const reducer = requestsReducer({ actionType, resetOn: [actionType], multiple: true });`

**Comment by klis87** [comment]
ahh, u reset on the request action itself

I think there is a simpler solution to your problem. Did you consider using `pending === 0 && <Feature />`.
Or even simpler, you could use https://github.com/klis87/redux-saga-requests/tree/master/packages/redux-saga-requests-react#requestcontainer which simplifies this.

You could show spinners while loading too.

Anyway, I guess your solved your actions this way for caching purposes right? I will try to work on this during this weekend, so if that is the case, you won't need to have it this way anymore.

**Comment by megawac** [comment]
So I'm actually loading features into a map via these requests so I'll load for example:

- streets
- places (restaurants, other pois)
- landmarks
- topography
- imagery

etc into a [map](https://www.openstreetmap.org/#map=16/43.4826/-80.4618) in parallel and as the assets load, I'll render them into view. I have a loading indicator at the bottom right of my screen saying some assets are still loading.

> Did you consider using pending === 0 && <Feature />

What do you mean by that?

**Comment by klis87** [comment]
> When I dispatch my request action type for each of those assets I want to reset my reducers states while the fetches are in progress and as the assets load they will appear in my interface. The reason I want to reset is that I want the assets to load into on demand but I don't want them to be visible for the previous map.

To rephrase, you just don't want to show `data` of your reducers while new requests are pending? `requestsReducer` keeps track of number of pending requests of a given type, it has `pending` integer next to `data`. So instead of resetting `data`, you can know that a new request is `pending` by looking at `pending` counter. If it is bigger that 0, it means that you have a pending request. So instead of rendering `data`, you can render nothing or a spinner if u prefer

**Comment by klis87** [comment]
Alternatively, this option could help you instead of resetting:
`onRequest: (state, action, config) => nextState`: here you can adjust how requestReducers handles request actions

**Comment by megawac** [comment]
I want my the majority of my views to be agnostic of the loading process, as far they care the data has changed so they will rerender with the new content, whether that content be the default state (empty feature collections) or data from the server. I'm slowly porting over a legacy app to a new system so I don't have time to replace everything, I'm using a redux inteface to drive the previous data hooks that the app was using as I port over the app. At the moment the pending state is not available at the view level without significant effort.

**Comment by megawac** [comment]
I have a loading indicator driven by the pending states of redux in the app, its just not coupled directly with the map views.

**Comment by klis87** [comment]
Ok, I see. But `onRequest` option of reducer should be perfect for resetting `data`?

**Comment by megawac** [comment]
Yeah, I could use `onRequest` myself I just don't see why I shouldn't be allowed to use `resetOn: [myActionType]` to accomplish the same thing

**Comment by klis87** [comment]
For me resetting on request itself  sound confusing and it complicates implementation.  I want to keep it as simple as possible, because soon I am going to work on kind of manager of `requestsReducers`, so actually it will be possible to stop writing `requestsReducers` completely. Apart from not writing reducers, big advantage will be that you wont need to write selectors for your data too, request action type will suffice.

Thats why I want to keep reducer as simple as possible. I will accept different kind of flag though, like `resetDataOnRequest` default as `false`, which will be used in `onRequest`:
```js
onRequest: (state, action, config) => ({
    ...state,
    pending: state.pending + 1,
    data: config.resetDataOnRequest ? config.getDefaultData(config.multiple) : state.data,
    error: null,
  }),
```

I know that probably your solution is cleaner, but my is encapsulated in `onRequest`, which doesnt affect `requestsReducer` body. I just need to think about it and later I might switch to your solution, but for now pls update PR to my implementation, I guess it will do exactly what you need.

**Comment by megawac** [comment]
Eh, lets sit on this then. I don't like that solution as much and it requires extra documentation.

**Comment by klis87** [comment]
Yeah, extra option is a serious disadvantage of my solution. Lets keep this issue and PR then, I will deal with it once working on new cetralized reducer. Dont worry though, `requestReducer` will still be supported.

**Comment by klis87** [comment]
@megawac I thought about it, and we will need to have it anyway in incoming reducer manager, so why not to add it now. Thx and sry I didn't merge it immediately, I just needed to make sure it will be in sync with new features.

published as `0.23.1`
